codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -361,9 +362,19 @@ public void error(ChannelHandlerContext ctx, Throwable cause) {
         @Override
         public void write(ChannelHandlerContext ctx, int allowedBytes) {
             int queuedData = queue.readableBytes();
-            if (!endOfStream && (queuedData == 0 || allowedBytes == 0)) {
-                // Nothing to write and we don't have to force a write because of EOS.
-                return;
+            if (!endOfStream) {
+                if (queuedData == 0) {
+                    // There's no need to write any data frames because there are only empty data frames in the queue
+                    // and it is not end of stream yet. Just complete their promises by writing an empty buffer.
+                    ChannelPromise writePromise = ctx.newPromise().addListener(this);
+                    queue.remove(0, writePromise).release();
+                    ctx.write(Unpooled.EMPTY_BUFFER, writePromise);
```

Comment:
I think there's no point of sending an empty data frame when EOS is false because .. it's nothing. Once thing I'm not sure about is the case where padding is not 0. Should we send an empty data frame with non-zero padding?